### PR TITLE
Report potential decommission recommission

### PR DESCRIPTION
**Pull request by Doy-lee**
Commit on top of: https://github.com/loki-project/loki/pull/841

```
    Report portential decomission and recommission for your node

    When processing a quorum for a block, if you are not in the quorum to
    validate other nodes, check if you're a node that is going to be tested.
    If you are, check based on your current data if you're potentially
    a candidate to be decommissioned/deregistered and if so report it to the
    console log.

    Note that this is only a heuristic and ultimately the decision lies on
    what the the other Service Nodes perceive the current state of your node
    is (i.e. if they're acting malicious then you will be deregistered
    irrespectively).
```

@jagerman

I'm not sure if this is desired, I found it a bit helpful for debugging on testnet, but it is just a heuristic on why you might be getting decomm/deregistered. Maybe it will lead to more questions than answers by other operators if it says one thing but your node gets affected otherwise.

Also not sure if my reorganizing of the process quorum loop made it more complicated or easier to understand.

It prints something like

```
Service node (yours) is active but is not passing tests for quorum: 100000
Service Node is passing all tests, if decommissioned, you will be recommissioned if status is maintained by the next quorum
```

or 

```
2019-09-17 06:57:26.140	I Temporary decommission for service node (yours): <b4d5d5336107fff5878d339f8de296e92f1014b071bb8377196d9afa36cd6277>
2019-09-17 06:59:24.322	I Submitted uptime-proof for service node (yours): <b4d5d5336107fff5878d339f8de296e92f1014b071bb8377196d9afa36cd6277>
2019-09-17 07:00:58.516	W Service Node (yours) is currently decommissioned and being tested in quorum: 129195
2019-09-17 07:00:58.516	W Service Node is currently failing the following tests: Uptime proof missing. Note: Storage server may not be reachable. This is only testable by an external Service Node.
2019-09-17 07:04:24.366	I Submitted uptime-proof for service node (yours): <b4d5d5336107fff5878d339f8de296e92f1014b071bb8377196d9afa36cd6277>

```
etcetera

Also there might be some potential with further networking improvements that we could contact the nodes in the quorum to figure out exactly why they're failing us.

### CODE REVIEW

**Comment by Doy-lee** [comment]
I think this is buggy. One sec on reviewing

**Comment by Doy-lee** [comment]
Passing tests now, should be good for review.

File: src/cryptonote_core/service_node_quorum_cop.cpp [hunk 1]
```diff
@@ -44,6 +44,34 @@
...
+  char const *service_node_test_results::why(bool testing_myself) const
+  {
+    static char buf[2048];
+    buf[0]              = 0;
+    char *buf_ptr       = buf;
+    char const *buf_end = buf + sizeof(buf);
+
+    if (passed())
+    {
+      buf_ptr += snprintf(buf_ptr, buf_end - buf_ptr, "Service Node is passing all tests, if decommissioned, you will be recommissioned if status is maintained by the next quorum");
```
Comment (jagerman) [inline src/cryptonote_core/service_node_quorum_cop.cpp]: I think we should shorten this to just "Service node is passing all tests".  Otherwise we're going to invite some edge case where someone sees the message but a dereg has already been sent to the mempool, or some test that can't be easily detected is failing, or ...

File: src/cryptonote_core/service_node_quorum_cop.cpp [hunk 2]
```diff
@@ -44,6 +44,34 @@
...
+    }
+    else
+    {
+      buf_ptr += snprintf(buf_ptr, buf_end - buf_ptr, "Service Node is currently failing the following tests: ");
+      if (!uptime_proved)         buf_ptr += snprintf(buf_ptr, buf_end - buf_ptr, "Uptime proof missing. ");
+      if (!voted_in_checkpoints)  buf_ptr += snprintf(buf_ptr, buf_end - buf_ptr, "Skipped voting in atleast %d checkpoints. ", (int)(CHECKPOINT_NUM_QUORUMS_TO_PARTICIPATE_IN - CHECKPOINT_MAX_MISSABLE_VOTES));
```
Comment (jagerman) [inline src/cryptonote_core/service_node_quorum_cop.cpp]: Missing space in "atleast"

File: src/cryptonote_core/service_node_quorum_cop.cpp [hunk 3]
```diff
@@ -212,33 +241,36 @@ namespace service_nodes
 
         case quorum_type::obligations:
         {
-          // NOTE: Wait atleast 2 hours before we're allowed to vote so that we collect necessary voting information
-          // from people on the network
-          time_t const now = time(nullptr);
+
+          // NOTE: Wait atleast 2 hours before we're allowed to vote so that we collect necessary voting information from people on the network
           bool alive_for_min_time = (now - m_core.get_start_time()) >= MIN_TIME_IN_S_BEFORE_VOTING;
```
Comment (jagerman) [inline src/cryptonote_core/service_node_quorum_cop.cpp]: `(now - m_core.get_start_time())` -> `live_time`

File: src/cryptonote_core/service_node_quorum_cop.h [hunk 4]
```diff
@@ -80,6 +80,7 @@ namespace service_nodes
     bool voted_in_checkpoints     = true;
     bool storage_server_reachable = true;
 
+    char const *why(bool testing_myself) const;
```
Comment (jagerman) [inline src/cryptonote_core/service_node_quorum_cop.h]: It looks like this is always called with `testing_myself` set to `true`

File: src/daemon/rpc_command_executor.cpp [hunk 5]
```diff
@@ -2398,32 +2398,42 @@ static void append_printable_service_node_list_entry(cryptonote::network_type ne
     buffer.append(indent1);
     buffer.append("[");
     buffer.append(std::to_string(entry_index));
-    buffer.append("] Service Node: ");
+    buffer.append("] ");
+    buffer.append("Service Node: ");
```
Comment (jagerman) [inline src/daemon/rpc_command_executor.cpp]: This is mainly just style, and I know this is just following the existing code, but all of these `buffer.append(blah);` would look so much nicer as `buffer += blah;`

Comment (jagerman) [reply src/daemon/rpc_command_executor.cpp]: ... or `std::ostringstream o; o << blah; ....; buffer += o.str();` which would also save a bunch of `std::to_string`s.

File: src/daemon/rpc_command_executor.cpp [hunk 6]
```diff
@@ -2490,23 +2502,61 @@ static void append_printable_service_node_list_entry(cryptonote::network_type ne
...
-      buffer.append("IP Address: ");
+    buffer.append("\n");
+    buffer.append(indent2);
+    buffer.append("IP Address & Port: ");
+    if (entry.public_ip == "0.0.0.0")
+    {
+        buffer.append("(Awaiting confirmation from network)");
+    }
+    else
+    {
```
Comment (jagerman) [inline src/daemon/rpc_command_executor.cpp]: I like changing the IP to "(Awaiting confirmation from network)", but I think these storage port should remain separate because we'll also have lokinet port & quorumnet port before long.

File: src/rpc/core_rpc_server_commands_defs.h [hunk 7]
```diff
@@ -2763,15 +2764,23 @@ namespace cryptonote
...
+        bool                                               storage_server_reachable;            // Whether the node's storage server has been reported as unreachable for a long time
+        uint64_t                                           storage_server_reachable_timestamp;  // The last time this Service Node's storage server was contacted
+        uint16_t                                           version_major;                       // Major version the node is currently running
+        uint16_t                                           version_minor;                       // Minor version the node is currently running
+        uint16_t                                           version_patch;                       // Patch version the node is currently running
+        std::vector<service_nodes::checkpoint_vote_record> votes;                               // Of the last N checkpoints the Service Node is in a checkpointing quorum, record whether or not the Service Node voted to checkpoint a block
```
Comment (jagerman) [inline src/rpc/core_rpc_server_commands_defs.h]: I think we need to deprecate this soon (though not saying in this PR) and stop adding to it and instead move everything to `get_n_service_nodes` so that callers pick-and-choose what they want and so we can stop mostly duplicating these two structures.  But mainly because the default `get_service_nodes` data response is getting *massive*.

File: src/daemon/rpc_command_executor.cpp (hunk 6 above)
Comment (jagerman) [reply src/daemon/rpc_command_executor.cpp]: Actually never mind, this is fine on one line.  Later we can change it to something like:

1.2.3.4 :23456 (storage-server), :22025 (qnet), :22026 (lokinet)

File: src/cryptonote_basic/hardfork.h [hunk 8]
```diff
@@ -47,6 +47,7 @@ namespace cryptonote
     };
 
     constexpr static uint8_t INVALID_HF_VERSION_FOR_HEIGHT = 255;
+    constexpr static uint64_t INVALID_HF_VERSION_HEIGHT    = -1;
```
[the remark this answers is not in the document]
Comment (Doy-lee) [reply src/cryptonote_basic/hardfork.h]: Added

File: src/cryptonote_core/service_node_quorum_cop.cpp (hunk 1 above)
Comment (Doy-lee) [reply src/cryptonote_core/service_node_quorum_cop.cpp]: Yes, good idea

File: src/cryptonote_core/service_node_quorum_cop.cpp (hunk 2 above)
Comment (Doy-lee) [reply src/cryptonote_core/service_node_quorum_cop.cpp]: Added

File: src/cryptonote_core/service_node_quorum_cop.cpp (hunk 3 above)
Comment (Doy-lee) [reply src/cryptonote_core/service_node_quorum_cop.cpp]: Added

File: src/cryptonote_core/service_node_quorum_cop.h (hunk 4 above)
Comment (Doy-lee) [reply src/cryptonote_core/service_node_quorum_cop.h]: Removed

File: src/daemon/rpc_command_executor.cpp (hunk 5 above)
Comment (Doy-lee) [reply src/daemon/rpc_command_executor.cpp]: Added

File: src/rpc/core_rpc_server_commands_defs.h (hunk 7 above)
Comment (Doy-lee) [reply src/rpc/core_rpc_server_commands_defs.h]: Agreed. We need to come up with some deprecating protocol for RPC. Various external services are using this endpoint.
